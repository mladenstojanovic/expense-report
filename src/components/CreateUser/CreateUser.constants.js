export const TYPE_ALL = 'all';
export const TYPE_PHONE = 'phone';
export const TYPE_EMAIL = 'email';
export const EMPTY_FORM_ERROR_MESSAGE =
  'Please enter your email or phone number';
export const INVALID_PHONE_AND_EMAIL_MESSAGE =
  'Email and phone number are invalid';
export const INVALID_PHONE_MESSAGE = 'Phone number is not valid';
export const INVALID_EMAIL_MESSAGE = 'Email is not valid';
