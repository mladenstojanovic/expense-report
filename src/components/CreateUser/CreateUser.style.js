import styled from 'styled-components/macro';

export const CreateUserStyle = styled.div``;

export const CreateUserFormStyle = styled.form`
  display: flex;
  flex-direction: column;
  padding-bottom: 1.25rem;
`;

export const CreateUserFormLabelStyle = styled.label`
  font-size: ${props => props.theme.fonts.sizes.heading};
  margin-bottom: 3px;
`;

export const CreateUserFormInputStyle = styled.input`
  width: 310px;
  height: 30px;
  margin-bottom: 10px;
  padding-left: 15px;
  border: 1px solid
    ${props => (props.error ? 'red' : props.theme.colors.lightBlue)};
  border-left: 4px solid
    ${props => (props.error ? 'red' : props.theme.colors.lightBlue)};
  border-radius: 5px;
  font-size: ${props => props.theme.fonts.sizes.paragraph};
`;

export const CreateUserFormInputButtonStyle = styled.input`
  height: 40px;
  margin-top: 20px;
  background: transparent;
  border: 1px solid ${props => props.theme.colors.lightBlue};
  border-radius: 5px;
  color: white;
  font-weight: bold;
  cursor: pointer;
  transition: border-color 0.3s ease-out, background 0.3s ease-out;

  &:hover {
    background: ${props => props.theme.colors.lightBlue};
    border-color: white;
  }
`;
