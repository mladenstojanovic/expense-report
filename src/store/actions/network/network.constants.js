export const GET_AUTH_TOKEN_START = '[NETWORK] Get Auth Token Start';
export const GET_AUTH_TOKEN_SUCCESS = '[NETWORK] Get Auth Token Success';
export const GET_AUTH_TOKEN_ERROR = '[NETWORK] Get Auth Token Error';
export const CREATE_USER_START = '[NETWORK] Create User Start';
export const CREATE_USER_SUCCESS = '[NETWORK] Create User Success';
export const CREATE_USER_ERROR = '[NETWORK] Create User Error';
export const ADD_BANK_CONNECTION_START = '[NETWORK] Add Bank Connection Start';
export const ADD_BANK_CONNECTION_SUCCESS =
  '[NETWORK] Add Bank Connection Success';
export const ADD_BANK_CONNECTION_ERROR = '[NETWORK] Add Bank Connection Error';
export const GET_TRANSACTIONS_START = '[NETWORK] Get Transactions Start';
export const GET_TRANSACTIONS_SUCCESS = '[NETWORK] Get Transactions Success';
export const GET_TRANSACTIONS_ERROR = '[NETWORK] Get Transactions Error';
export const NETWORK_OPERATIONS_START = '[NETWORK] Network Operations Start';
export const CONNECTION_JOB_START = '[NETWORK] Connection Job Start';
export const CONNECTION_JOB_SUCCESS = '[NETWORK] Connection Job Success';
export const CONNECTION_JOB_ERROR = '[NETWORK] Connection Job Error';

export const IN_PROGRESS = 'in progress';
export const SUCCESS = 'success';
export const ERROR = 'error';

export const RETRIEVE_TRANSACTIONS = 'retrieve-transactions';
