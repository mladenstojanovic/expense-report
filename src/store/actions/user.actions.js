import Axios from 'axios';
const GET_AUTH_TOKEN_START = '[AUTH] Get Auth Token Start';

export const getAuthTokenStart = () => ({
  type: GET_AUTH_TOKEN_START
});

export const submitUser = userData => {
  return async dispatch => {
    try {
      dispatch(getAuthTokenStart());

      console.log('USER DATAAAAAAAAA', userData);
      console.log('aaaaaaaaaaaaaaaa', `Basic ${process.env.REACT_APP_API_KEY}`);
      //   const header = `Basic ${process.env.REACT_APP_API_KEY}`;
      //   const authResponse = await Axios.post(
      //     'https://au-api.basiq.io/token',
      //     {},
      //     {
      //       headers: {
      //         Authorization: `Basic ${process.env.REACT_APP_API_KEY}`,
      //         'Content-Type': 'application/x-www-form-urlencoded',
      //         'basiq-version': '2.0'
      //       }
      //     }
      //   );
      const scope = new URLSearchParams(
        'https://au-api.basiq.io/token?scope=CLIENT_ACCESS'
      );
      const authResponse = await fetch(
        'https://au-api.basiq.io/token?scope=CLIENT_ACCESS',
        {
          method: 'POST',
          headers: {
            Authorization: `Basic ${process.env.REACT_APP_API_KEY}`,
            'Content-Type': 'application/x-www-form-urlencoded',
            'basiq-version': '2.0'
          },
          body: scope
        }
      );
      console.log('AUTH RESPONSE', authResponse);
    } catch (err) {
      console.log('ERRORRRRRRRRRRRR', err);
    }
  };
};
