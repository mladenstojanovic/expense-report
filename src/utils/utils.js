import React from 'react';
import { render } from '@testing-library/react';
import { createStore, combineReducers } from 'redux';
import { ThemeProvider } from 'styled-components';
import { Provider } from 'react-redux';
import networkReducer from '../store/reducers/network.reducer';
import theme from '../styles/theme';

export const sleep = milliseconds => {
  return new Promise(resolve => setTimeout(resolve, milliseconds));
};

export const sortTransactionData = data => {
  const sortedData = {};
  for (let index = 0; index < data.length; index++) {
    if (!data[index].subClass) {
      continue;
    }
    if (!sortedData[data[index].subClass.code]) {
      sortedData[data[index].subClass.code] = {
        total: 0,
        numOfTransactions: 0,
        title: data[index].subClass.title
      };
    }
    sortedData[data[index].subClass.code].total += Math.abs(
      parseInt(data[index].amount)
    );
    sortedData[data[index].subClass.code].numOfTransactions++;
  }
  return sortedData;
};

export const isEmailValid = email => /^[^\s@]+@[^\s@]+\.[^\s@]+$/.test(email);

export const isPhoneNumberValid = phone =>
  /^[+]?[(]?[0-9]{3}[)]?[-\s.]?[0-9]{3}[-\s.]?[0-9]{4,6}$/.test(phone);

export const renderWithReduxAndStyles = (
  ui,
  {
    initialState,
    store = createStore(
      combineReducers({ network: networkReducer }),
      initialState
    )
  } = {}
) => ({
  ...render(
    <Provider store={store}>
      <ThemeProvider theme={theme}>{ui}</ThemeProvider>
    </Provider>
  ),
  store
});
