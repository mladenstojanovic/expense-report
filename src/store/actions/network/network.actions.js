import {
  GET_AUTH_TOKEN_START,
  CREATE_USER_START,
  ADD_BANK_CONNECTION_START,
  GET_TRANSACTIONS_START,
  GET_AUTH_TOKEN_SUCCESS,
  CREATE_USER_SUCCESS,
  ADD_BANK_CONNECTION_SUCCESS,
  GET_TRANSACTIONS_SUCCESS,
  NETWORK_OPERATIONS_START,
  CONNECTION_JOB_START,
  CONNECTION_JOB_SUCCESS,
  GET_AUTH_TOKEN_ERROR,
  CREATE_USER_ERROR,
  ADD_BANK_CONNECTION_ERROR,
  CONNECTION_JOB_ERROR,
  GET_TRANSACTIONS_ERROR,
  SUCCESS,
  RETRIEVE_TRANSACTIONS
} from './network.constants';
import { sleep, sortTransactionData } from '../../../utils/utils';

export const getAuthTokenStart = () => ({
  type: GET_AUTH_TOKEN_START
});

export const getAuthTokenSuccess = () => ({
  type: GET_AUTH_TOKEN_SUCCESS
});

export const getAuthTokenError = () => ({
  type: GET_AUTH_TOKEN_ERROR
});

export const createUserStart = () => ({
  type: CREATE_USER_START
});

export const createUserSuccess = () => ({
  type: CREATE_USER_SUCCESS
});

export const createUserError = () => ({
  type: CREATE_USER_ERROR
});

export const addBankConnectionStart = () => ({
  type: ADD_BANK_CONNECTION_START
});

export const addBankConnectionSuccess = () => ({
  type: ADD_BANK_CONNECTION_SUCCESS
});

export const addBankConnectionError = () => ({
  type: ADD_BANK_CONNECTION_ERROR
});

export const connectionJobStart = () => ({
  type: CONNECTION_JOB_START
});

export const connectionJobSuccess = () => ({
  type: CONNECTION_JOB_SUCCESS
});

export const connectionJobError = () => ({
  type: CONNECTION_JOB_ERROR
});

export const getTransactionsStart = () => ({
  type: GET_TRANSACTIONS_START
});

export const getTransactionsSuccess = payload => ({
  type: GET_TRANSACTIONS_SUCCESS,
  payload
});

export const getTransactionsError = () => ({
  type: GET_TRANSACTIONS_ERROR
});

export const networkOperationsStart = () => ({
  type: NETWORK_OPERATIONS_START
});

let userId;
let token;
let parsedAddConnectionResponse;
let jobFinished = false;

export const getAuthToken = async dispatch => {
  try {
    dispatch(networkOperationsStart());
    dispatch(getAuthTokenStart());

    const authResponse = await fetch(process.env.REACT_APP_TOKEN_ENDPOINT, {
      headers: {
        Authorization: `Basic ${process.env.REACT_APP_API_KEY}`
      }
    });
    const authResponseParsed = await authResponse.json();

    token = authResponseParsed.access_token;
    dispatch(getAuthTokenSuccess());
  } catch {
    dispatch(getAuthTokenError());
  }
};

export const createUser = async (dispatch, userData) => {
  try {
    dispatch(createUserStart());

    const createUserResponse = await fetch(
      `${process.env.REACT_APP_BASE_ENDPOINT}/users`,
      {
        method: 'POST',
        body: JSON.stringify(userData),
        headers: {
          Authorization: `Bearer ${token}`
        }
      }
    );
    const createUserResponseJson = await createUserResponse.json();

    userId = createUserResponseJson.id;
    dispatch(createUserSuccess());
  } catch {
    dispatch(createUserError());
  }
};

export const addConnection = async dispatch => {
  try {
    const testUser = {
      loginId: process.env.REACT_APP_LOGIN_ID,
      password: process.env.REACT_APP_LOGIN_PWD,
      institution: {
        id: process.env.REACT_APP_LOGIN_INSTITUTION_ID
      }
    };

    dispatch(addBankConnectionStart());

    const addConnectionResponse = await fetch(
      `${process.env.REACT_APP_BASE_ENDPOINT}/users/${userId}/connections`,
      {
        method: 'POST',
        body: JSON.stringify(testUser),
        headers: {
          Authorization: `Bearer ${token}`,
          'content-type': 'application/json'
        }
      }
    );
    parsedAddConnectionResponse = await addConnectionResponse.json();

    // mozda neki error handling?

    dispatch(addBankConnectionSuccess());
  } catch {
    dispatch(addBankConnectionError());
  }
};

export const connectionJob = async dispatch => {
  try {
    dispatch(connectionJobStart());
    //@TODO: config?
    let retries = 20;
    do {
      //@TODO: config?
      await sleep(10000);

      const job = await fetch(`${parsedAddConnectionResponse.links.self}`, {
        headers: {
          Authorization: `Bearer ${token}`,
          'content-type': 'application/json'
        }
      });
      const parsedJob = await job.json();
      const retrieveTransactionJob = parsedJob.steps.find(
        job => job.title === RETRIEVE_TRANSACTIONS
      );

      if (retrieveTransactionJob.status === SUCCESS) {
        jobFinished = true;
      } else {
        retries--;
      }
    } while (!jobFinished && retries > 0);
    if (jobFinished) {
      dispatch(connectionJobSuccess());
    } else {
      dispatch(connectionJobError());
    }
  } catch {
    dispatch(connectionJobError());
  }
};

export const getTransactions = async dispatch => {
  try {
    dispatch(getTransactionsStart());
    const getTransactionsResponse = await fetch(
      `${process.env.REACT_APP_BASE_ENDPOINT}/users/${userId}/transactions`,
      {
        headers: {
          Authorization: `Bearer ${token}`
        }
      }
    );
    const parsedGetTransactionsResponse = await getTransactionsResponse.json();

    dispatch(
      getTransactionsSuccess(
        sortTransactionData(parsedGetTransactionsResponse.data)
      )
    );
  } catch {
    dispatch(getTransactionsError());
  }
};

export const submitUser = userData => {
  return async dispatch => {
    await getAuthToken(dispatch);
    await createUser(dispatch, userData);
    await addConnection(dispatch);
    await connectionJob(dispatch);
    if (jobFinished) {
      await getTransactions(dispatch);
    }
  };
};
