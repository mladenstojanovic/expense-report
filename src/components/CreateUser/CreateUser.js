import React, { useState } from 'react';
import { useDispatch } from 'react-redux';
import { submitUser } from '../../store/actions/network/network.actions';
import {
  CreateUserStyle,
  CreateUserFormInputButtonStyle,
  CreateUserFormStyle,
  CreateUserFormInputStyle,
  CreateUserFormLabelStyle
} from './CreateUser.style';
import { isPhoneNumberValid, isEmailValid } from '../../utils/utils';
import {
  TYPE_ALL,
  EMPTY_FORM_ERROR_MESSAGE,
  INVALID_PHONE_AND_EMAIL_MESSAGE,
  TYPE_PHONE,
  INVALID_PHONE_MESSAGE,
  TYPE_EMAIL,
  INVALID_EMAIL_MESSAGE
} from './CreateUser.constants';

const defaultErrorState = {
  type: '',
  errorText: ''
};

export const handleSubmit = (event, email, phone, setError, dispatch) => {
  event.preventDefault();
  if (!email && !phone) {
    setError({
      type: TYPE_ALL,
      errorText: EMPTY_FORM_ERROR_MESSAGE
    });
    return;
  }
  if (email && phone && !isPhoneNumberValid(phone) && !isEmailValid(email)) {
    setError({
      type: TYPE_ALL,
      errorText: INVALID_PHONE_AND_EMAIL_MESSAGE
    });
    return;
  }
  if (phone && !isPhoneNumberValid(phone)) {
    setError({
      type: TYPE_PHONE,
      errorText: INVALID_PHONE_MESSAGE
    });
    return;
  }
  if (email && !isEmailValid(email)) {
    setError({ type: TYPE_EMAIL, errorText: INVALID_EMAIL_MESSAGE });
    return;
  }
  dispatch(submitUser({ email, phone }));
};

const CreateUser = () => {
  const [email, setEmail] = useState('');
  const [phone, setPhone] = useState('');
  const [error, setError] = useState(defaultErrorState);
  const dispatch = useDispatch();
  return (
    <>
      <CreateUserStyle>
        <CreateUserFormStyle
          onSubmit={event => {
            handleSubmit(event, email, phone, setError, dispatch);
          }}
        >
          <CreateUserFormLabelStyle htmlFor="email">
            Email
          </CreateUserFormLabelStyle>
          <CreateUserFormInputStyle
            id="email"
            type="email"
            name="email"
            data-testid="emailInput"
            value={email}
            error={error.type === 'email' || error.type === 'all'}
            onChange={event => {
              if (error.errorText) {
                setError(defaultErrorState);
              }
              setEmail(event.target.value);
            }}
          />
          <CreateUserFormLabelStyle htmlFor="phone">
            Phone Number
          </CreateUserFormLabelStyle>
          <CreateUserFormInputStyle
            id="phone"
            type="tel"
            name="phone"
            data-testid="phoneInput"
            value={phone}
            error={error.type === 'phone' || error.type === 'all'}
            onChange={event => {
              if (error.errorText) {
                setError(defaultErrorState);
              }
              setPhone(event.target.value);
            }}
          />
          <CreateUserFormInputButtonStyle
            type="submit"
            value="Submit"
            data-testid="formSubmit"
          />
        </CreateUserFormStyle>
      </CreateUserStyle>
      {error.errorText && <p data-testid="errorMessage">{error.errorText}</p>}
    </>
  );
};

export default CreateUser;
