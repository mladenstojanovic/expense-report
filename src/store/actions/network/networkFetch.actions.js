import {
  networkOperationsStart,
  getAuthTokenStart,
  getAuthTokenError,
  createUserStart,
  createUserError,
  createUserSuccess,
  addBankConnectionStart,
  addBankConnectionError,
  addBankConnectionSuccess,
  connectionJobStart,
  connectionJobSuccess,
  connectionJobError,
  getTransactionsStart,
  getTransactionsError,
  getAuthTokenSuccess,
  getTransactionsSuccess
} from './network.actions';
import { sleep, sortTransactionData } from '../../../utils/utils';
import { RETRIEVE_TRANSACTIONS, SUCCESS } from './network.constants';

let userId;
let token;
let parsedAddConnectionResponse;
let jobFinished = false;

export const getAuthToken = async dispatch => {
  try {
    dispatch(networkOperationsStart());
    dispatch(getAuthTokenStart());

    const authResponse = await fetch(process.env.REACT_APP_TOKEN_ENDPOINT, {
      headers: {
        Authorization: `Basic ${process.env.REACT_APP_API_KEY}`
      }
    });

    const authResponseParsed = await authResponse.json();

    token = authResponseParsed.access_token;

    if (!token) {
      dispatch(getAuthTokenError());
      return {
        error: true
      };
    }

    dispatch(getAuthTokenSuccess());
    return {
      error: false
    };
  } catch {
    dispatch(getAuthTokenError());
    return {
      error: true
    };
  }
};

export const createUser = async (dispatch, userData) => {
  try {
    dispatch(createUserStart());

    const createUserResponse = await fetch(
      `${process.env.REACT_APP_BASE_ENDPOINT}/users`,
      {
        method: 'POST',
        body: JSON.stringify(userData),
        headers: {
          Authorization: `Bearer ${token}`
        }
      }
    );
    const createUserResponseJson = await createUserResponse.json();

    userId = createUserResponseJson.id;

    if (!userId) {
      dispatch(createUserError());
      return {
        error: true
      };
    }
    dispatch(createUserSuccess());
    return {
      error: false
    };
  } catch {
    dispatch(createUserError());
    return {
      error: true
    };
  }
};

export const addConnection = async dispatch => {
  try {
    const testUser = {
      loginId: process.env.REACT_APP_LOGIN_ID,
      password: process.env.REACT_APP_LOGIN_PWD,
      institution: {
        id: process.env.REACT_APP_LOGIN_INSTITUTION_ID
      }
    };

    dispatch(addBankConnectionStart());

    const addConnectionResponse = await fetch(
      `${process.env.REACT_APP_BASE_ENDPOINT}/users/${userId}/connections`,
      {
        method: 'POST',
        body: JSON.stringify(testUser),
        headers: {
          Authorization: `Bearer ${token}`,
          'content-type': 'application/json'
        }
      }
    );
    parsedAddConnectionResponse = await addConnectionResponse.json();

    const jobId = parsedAddConnectionResponse.id;

    if (!jobId) {
      dispatch(addBankConnectionError());
      return {
        error: true
      };
    }

    dispatch(addBankConnectionSuccess());
    return { error: false, parsedAddConnectionResponse };
  } catch {
    dispatch(addBankConnectionError());
    return {
      error: true
    };
  }
};

export const connectionJob = async (
  dispatch,
  pollingRetries,
  pollingTime,
  parsedAddConnectionResponse
) => {
  try {
    dispatch(connectionJobStart());
    let retries = pollingRetries;
    do {
      await sleep(pollingTime);

      const job = await fetch(`${parsedAddConnectionResponse.links.self}`, {
        headers: {
          Authorization: `Bearer ${token}`,
          'content-type': 'application/json'
        }
      });
      const parsedJob = await job.json();
      const retrieveTransactionJob = parsedJob.steps.find(
        job => job.title === RETRIEVE_TRANSACTIONS
      );

      if (retrieveTransactionJob.status === SUCCESS) {
        jobFinished = true;
      } else {
        retries--;
      }
    } while (!jobFinished && retries > 0);
    if (jobFinished) {
      dispatch(connectionJobSuccess());
      return true;
    } else {
      dispatch(connectionJobError());
      return false;
    }
  } catch {
    dispatch(connectionJobError());
    return false;
  }
};

export const getTransactions = async dispatch => {
  try {
    dispatch(getTransactionsStart());
    const getTransactionsResponse = await fetch(
      `${process.env.REACT_APP_BASE_ENDPOINT}/users/${userId}/transactions`,
      {
        headers: {
          Authorization: `Bearer ${token}`
        }
      }
    );
    const parsedGetTransactionsResponse = await getTransactionsResponse.json();

    if (!parsedGetTransactionsResponse.data) {
      dispatch(getTransactionsError());
      return;
    }

    dispatch(
      getTransactionsSuccess(
        sortTransactionData(parsedGetTransactionsResponse.data)
      )
    );
  } catch {
    dispatch(getTransactionsError());
  }
};
