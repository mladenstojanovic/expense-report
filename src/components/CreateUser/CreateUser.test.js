import React from 'react';
import CreateUser from './CreateUser';
import { fireEvent } from '@testing-library/react';
import { renderWithReduxAndStyles } from '../../utils/utils';
import {
  EMPTY_FORM_ERROR_MESSAGE,
  INVALID_PHONE_AND_EMAIL_MESSAGE,
  INVALID_PHONE_MESSAGE,
  INVALID_EMAIL_MESSAGE
} from './CreateUser.constants';
import theme from '../../styles/theme';
import { submitUser } from '../../store/actions/network/network.actions';

jest.mock('../../store/actions/network/network.actions', () => ({
  submitUser: jest.fn().mockReturnValue({ type: 'TEST' })
}));

describe('Test CreateUser component', () => {
  it('should render correctly', () => {
    const { container } = renderWithReduxAndStyles(<CreateUser />);

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should display error message when empty form is submitted', () => {
    const { getByTestId } = renderWithReduxAndStyles(<CreateUser />);

    const emailInput = getByTestId('emailInput');
    const phoneInput = getByTestId('phoneInput');
    const submitButton = getByTestId('formSubmit');

    expect(emailInput).toHaveStyleRule(
      'border',
      `1px solid ${theme.colors.lightBlue}`
    );
    expect(phoneInput).toHaveStyleRule(
      'border',
      `1px solid ${theme.colors.lightBlue}`
    );

    fireEvent.click(submitButton);

    const errorMessage = getByTestId('errorMessage');

    expect(emailInput).toHaveStyleRule('border', '1px solid red');
    expect(phoneInput).toHaveStyleRule('border', '1px solid red');
    expect(errorMessage).toBeInTheDocument();
    expect(errorMessage).toHaveTextContent(EMPTY_FORM_ERROR_MESSAGE);
  });

  it('should display error message when both email and phone number are invalid', () => {
    const { getByTestId } = renderWithReduxAndStyles(<CreateUser />);

    const emailInput = getByTestId('emailInput');
    const phoneInput = getByTestId('phoneInput');
    const submitButton = getByTestId('formSubmit');

    expect(emailInput).toHaveStyleRule(
      'border',
      `1px solid ${theme.colors.lightBlue}`
    );
    expect(phoneInput).toHaveStyleRule(
      'border',
      `1px solid ${theme.colors.lightBlue}`
    );

    fireEvent.change(emailInput, { target: { value: 'bad@email' } });
    fireEvent.change(phoneInput, { target: { value: 'bad phone number' } });
    fireEvent.click(submitButton);

    const errorMessage = getByTestId('errorMessage');

    expect(emailInput).toHaveStyleRule('border', '1px solid red');
    expect(phoneInput).toHaveStyleRule('border', '1px solid red');
    expect(errorMessage).toBeInTheDocument();
    expect(errorMessage).toHaveTextContent(INVALID_PHONE_AND_EMAIL_MESSAGE);
  });

  it('should display error message when phone number is invalid', () => {
    const { getByTestId } = renderWithReduxAndStyles(<CreateUser />);

    const emailInput = getByTestId('emailInput');
    const phoneInput = getByTestId('phoneInput');
    const submitButton = getByTestId('formSubmit');

    expect(emailInput).toHaveStyleRule(
      'border',
      `1px solid ${theme.colors.lightBlue}`
    );
    expect(phoneInput).toHaveStyleRule(
      'border',
      `1px solid ${theme.colors.lightBlue}`
    );

    fireEvent.change(phoneInput, { target: { value: 'bad phone number' } });
    fireEvent.click(submitButton);

    const errorMessage = getByTestId('errorMessage');

    expect(emailInput).toHaveStyleRule(
      'border',
      `1px solid ${theme.colors.lightBlue}`
    );
    expect(phoneInput).toHaveStyleRule('border', '1px solid red');
    expect(errorMessage).toBeInTheDocument();
    expect(errorMessage).toHaveTextContent(INVALID_PHONE_MESSAGE);
  });

  it('should display error message when email is invalid', () => {
    const { getByTestId } = renderWithReduxAndStyles(<CreateUser />);

    const emailInput = getByTestId('emailInput');
    const phoneInput = getByTestId('phoneInput');
    const submitButton = getByTestId('formSubmit');

    expect(emailInput).toHaveStyleRule(
      'border',
      `1px solid ${theme.colors.lightBlue}`
    );
    expect(phoneInput).toHaveStyleRule(
      'border',
      `1px solid ${theme.colors.lightBlue}`
    );

    fireEvent.change(emailInput, { target: { value: 'bad@email' } });
    fireEvent.click(submitButton);

    const errorMessage = getByTestId('errorMessage');

    expect(phoneInput).toHaveStyleRule(
      'border',
      `1px solid ${theme.colors.lightBlue}`
    );
    expect(emailInput).toHaveStyleRule('border', '1px solid red');
    expect(errorMessage).toBeInTheDocument();
    expect(errorMessage).toHaveTextContent(INVALID_EMAIL_MESSAGE);
  });

  it('should clear the error message when you start typing again after getting an error', () => {
    const { getByTestId } = renderWithReduxAndStyles(<CreateUser />);

    const emailInput = getByTestId('emailInput');
    const phoneInput = getByTestId('phoneInput');
    const submitButton = getByTestId('formSubmit');

    expect(emailInput).toHaveStyleRule(
      'border',
      `1px solid ${theme.colors.lightBlue}`
    );
    expect(phoneInput).toHaveStyleRule(
      'border',
      `1px solid ${theme.colors.lightBlue}`
    );

    fireEvent.click(submitButton);

    const errorMessage = getByTestId('errorMessage');

    expect(phoneInput).toHaveStyleRule('border', '1px solid red');
    expect(emailInput).toHaveStyleRule('border', '1px solid red');
    expect(errorMessage).toBeInTheDocument();

    fireEvent.change(phoneInput, { target: { value: 'typing' } });

    expect(phoneInput).toHaveStyleRule(
      'border',
      `1px solid ${theme.colors.lightBlue}`
    );
    expect(emailInput).toHaveStyleRule(
      'border',
      `1px solid ${theme.colors.lightBlue}`
    );
    expect(errorMessage).not.toBeInTheDocument();

    fireEvent.click(submitButton);

    const newErrorMessage = getByTestId('errorMessage');

    expect(phoneInput).toHaveStyleRule('border', '1px solid red');
    expect(newErrorMessage).toBeInTheDocument();

    fireEvent.change(emailInput, { target: { value: 'typing' } });

    expect(phoneInput).toHaveStyleRule(
      'border',
      `1px solid ${theme.colors.lightBlue}`
    );
    expect(emailInput).toHaveStyleRule(
      'border',
      `1px solid ${theme.colors.lightBlue}`
    );
    expect(newErrorMessage).not.toBeInTheDocument();
  });

  it('should dispatch an action when all data is valid', () => {
    const { getByTestId } = renderWithReduxAndStyles(<CreateUser />);

    const emailInput = getByTestId('emailInput');
    const phoneInput = getByTestId('phoneInput');
    const submitButton = getByTestId('formSubmit');

    fireEvent.change(emailInput, { target: { value: 'good@email.com' } });
    fireEvent.change(phoneInput, { target: { value: '+381641234567' } });
    fireEvent.click(submitButton);

    expect(emailInput).toHaveStyleRule(
      'border',
      `1px solid ${theme.colors.lightBlue}`
    );
    expect(phoneInput).toHaveStyleRule(
      'border',
      `1px solid ${theme.colors.lightBlue}`
    );
    expect(submitUser).toHaveBeenCalledWith({
      email: 'good@email.com',
      phone: '+381641234567'
    });
  });
});
