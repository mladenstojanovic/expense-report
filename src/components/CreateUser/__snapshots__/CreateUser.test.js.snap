// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test CreateUser component should render correctly 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  padding-bottom: 1.25rem;
}

.c1 {
  margin-bottom: 3px;
}

.c2 {
  width: 310px;
  height: 30px;
  margin-bottom: 10px;
  padding-left: 15px;
  border: 1px solid #09d3ac;
  border-left: 4px solid #09d3ac;
  border-radius: 5px;
  font-size: 20px;
}

.c3 {
  height: 40px;
  margin-top: 20px;
  background: transparent;
  border: 1px solid #09d3ac;
  border-radius: 5px;
  color: white;
  font-weight: bold;
  cursor: pointer;
  -webkit-transition: border-color 0.3s ease-out,background 0.3s ease-out;
  transition: border-color 0.3s ease-out,background 0.3s ease-out;
}

.c3:hover {
  background: #09d3ac;
  border-color: white;
}

<div
  class=""
>
  <form
    class="c0"
  >
    <label
      class="c1"
      for="email"
    >
      Email
    </label>
    <input
      class="c2"
      data-testid="emailInput"
      id="email"
      name="email"
      type="email"
      value=""
    />
    <label
      class="c1"
      for="phone"
    >
      Phone Number
    </label>
    <input
      class="c2"
      data-testid="phoneInput"
      id="phone"
      name="phone"
      type="tel"
      value=""
    />
    <input
      class="c3"
      data-testid="formSubmit"
      type="submit"
      value="Submit"
    />
  </form>
</div>
`;
